### ISSUE TITLE
Parser output inconsistencies

### ISSUE THREAD

**Issue by kevinrenskers** [raml-org/raml-js-parser-2#453]
This is the JSON representation of parsing the [world-music-api](https://github.com/raml-org/raml-examples/tree/master/others/world-music-api) example.

```
{ annotations: { info: { structuredValue: [Object], name: 'info' } },
  uses: 
   [ { key: 'SongsLib', value: 'libraries/songs-library.raml' },
     { key: 'ApiLib', value: 'libraries/api-library.raml' } ],
  types: 
   [ { Entry: [Object] },
     { AnotherEntry: [Object] },
     { User: [Object] } ],
  traits: [ { secured: [Object] } ],
  resourceTypes: [ { collection: [Object] } ],
  annotationTypes: 
   [ { deprecated: [Object] },
     { monitoringInterval: [Object] },
     { ready: [Object] },
     { info: [Object] },
     { rediractable: [Object] } ],
  securitySchemes: 
   [ { oauth_1_0: [Object] },
     { oauth_2_0: [Object] },
     { custom_scheme: [Object] } ],
  title: 'World Music API',
  description: 'This is an example of a music API.',
  version: 'v1',
  baseUri: 'http://{environment}.musicapi.com/{version}',
  baseUriParameters: 
   { environment: 
      { name: 'environment',
        displayName: 'environment',
        type: [Object],
        required: true,
        enum: [Object],
        __METADATA__: [Object] },
     version: 
      { name: 'version',
        displayName: 'version',
        type: [Object],
        required: true,
        enum: [Object],
        __METADATA__: [Object] } },
  protocols: [ 'HTTP', 'HTTPS' ],
  mediaType: 'application/json',
  securedBy: [ 'custom_scheme' ],
  resources: 
   [ { methods: [Object],
       securedBy: [Object],
       relativeUri: '/api',
       displayName: '/api',
       __METADATA__: [Object],
       relativeUriPathSegments: [Object] },
     { methods: [Object],
       type: 'collection',
       securedBy: [Object],
       relativeUri: '/entry',
       displayName: '/entry',
       __METADATA__: [Object],
       relativeUriPathSegments: [Object] },
     { annotations: [Object],
       methods: [Object],
       is: [Object],
       description: 'Access to all songs inside the music world library.',
       securedBy: [Object],
       relativeUri: '/songs',
       displayName: 'Songs',
       resources: [Object],
       __METADATA__: [Object],
       relativeUriPathSegments: [Object] } ],
  documentation: 
   [ { title: 'Getting Started',
       content: 'This is a getting started guide for the World Music API.\n' },
     { title: 'Legal', content: 'See http://legal.musicapi.com' } ],
  scalarsAnnotations: { baseUri: [ [Object] ] } }
```

Some observations that I have:
#### 1. `types`, `traits`, `resourceTypes`, `annotationTypes`, `securitySchemes` are all an array of objects, where that object has a key and then the value is yet another object. For example:

```
securitySchemes:  [
    { oauth_1_0: [Object] },
    { oauth_2_0: [Object] },
    { custom_scheme: [Object] } 
],
```

I don't understand why this isn't just an object with 3 keys? 

```
securitySchemes: {
   oauth_1_0: [Object],
   oauth_2_0: [Object],
   custom_scheme: [Object] 
}
```
#### 3. `uses` is yet another structure: an array of objects, where each object is like a key/value database. Why is this completely different then anything else? Wouldn't this just work?

```
uses: {
  SongsLib: 'libraries/songs-library.raml',
  ApiLib: 'libraries/api-library.raml'
}
```
#### 4. Then there are things like resources which is an array of resource objects (awesome), and documentation which is an array of doc objects (awesome). But then the methods of a resource is an array like this:

```
[ { queryString: 
     { name: 'queryString',
       displayName: 'queryString',
       type: [Object],
       required: true,
       properties: [Object],
       __METADATA__: [Object] },
    protocols: [ 'HTTP' ],
    securedBy: [ 'custom_scheme' ],
    method: 'get',
    __METADATA__: { primitiveValuesMeta: [Object] } },
  { body: { 'application/json': [Object] },
    protocols: [ 'HTTP' ],
    securedBy: [ 'custom_scheme' ],
    method: 'post',
    __METADATA__: { primitiveValuesMeta: [Object] } } ]
```

So in here `body` is an object with the key `application/json` and then the value is that body (with a field `name` of `application/json`. So why isn't `body` just an array of body objects, instead of an object that contains objects?

Basically the problem with the output is that it's very inconsistent. The way how some things are arrays of plain objects, other things are an array of objects that hold objects, other things are an object that hold objects.. it's very hard to use this output in something like [raml2html](https://github.com/raml2html/raml2html).

I'm not sure if it's too late to do anything to make this more consistent so at the moment I am doing all I can to make it consistent within [raml2obj](https://github.com/raml2html/raml2obj) to simplify the raml2html and raml2md clients, but it would probably be a good long term improvement for this parser itself.

**Comment by sichvoge** [comment]
I completely agree here. The output needs to be consistent. There is another issue that is basically related to that: https://github.com/raml-org/raml-js-parser-2/issues/291. Especially [this comment](https://github.com/raml-org/raml-js-parser-2/issues/291#issuecomment-227742752) that illustrates different examples. Can you give some feedback here. We should consolidate both into this issue, since it speaks about the inconsistency problem more generally.

**Comment by sichvoge** [comment]
BTW, a lot decision has been driven by the fact to make the parser output more backwards compatible with the old parser.

**Comment by domarmstrong** [comment]
Does the parser really need to maintain backwards compatibility with the old parser? Who is this compatibility maintained for? Surely 1.0 is a good point to move to a better model (although passed that point now). It is a very counter intuitive model in a lot of places. Arrays of [{ key1: {}}, { key2: {}}] are really nasty.

**Comment by sichvoge** [comment]
I completely agree. It's not that we don't hear all of you ;) We will plan to remove the inconsistency with our planned refactoring.

**Comment by KonstantinSviridov** [comment]
Hi everyone!

We can not modify the old JSON format as we want to prevent our users from compatibility failure.

Your demands have been considered in the new JOSN format design. In comparison with the old format the new one has most maps replaced by arrays in a natural way.

The new format JSON is available by means of `load` and `loadSync` methods. Please, refer to the [getting started guide](https://github.com/raml-org/raml-js-parser-2/blob/master/documentation/GettingStartedNewApi.md) for details.

Regards,
Konstantin
